pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -689,27 +694,88 @@ func CredentialHelpers(sys *types.SystemContext) ([]string, error) {
 	return config.partialV2.CredentialHelpers, nil
 }
 
-// refMatchesPrefix returns true iff ref,
+// refMatchingSubdomainPrefix returns the length of ref without the repo name
+// iff ref, which is a registry, repository namespace, repository or image reference (as formatted by
+// reference.Domain(), reference.Named.Name() or reference.Reference.String()
+// — note that this requires the name to start with an explicit hostname!),
+// matches a Registry.Prefix value containing wildcarded subdomains in the
+// format: *.example.com. Wildcards are only accepted at the beginning, so
+// other formats like example.*.com will not work. Wildcarded prefixes also
+// cannot contain port numbers or namespaces in them.
+func refMatchingSubdomainPrefix(ref, prefix string) int {
+	origRef := ref
+
+	if strings.Contains(prefix[2:], "/") || strings.Contains(prefix[2:], ":") || strings.Contains(prefix[2:], "@") {
+		panic("Invalid wildcarded prefix")
+	}
+
+	if strings.LastIndex(ref, "/") == -1 {
+		return -1
+	}
+
+	index := strings.Index(ref, prefix[1:])
+	if index == -1 {
+		return -1
+	}
+
+	firstSlash := strings.Index(ref, "/")
+	if firstSlash != -1 {
+		ref = ref[:firstSlash]
+	}
+
+	if strings.Contains(ref, "@") || strings.Contains(ref[:index], ":") {
+		return -1
+	}
+
+	// Strip port number if any, before appending to res slice.
+	// Currently, the most compatible behavior is to return
+	// example.com:8443/ns, example.com:8443, *.com.
+	// If a port number is not specified, the expected behavior would be
+	// example.com/ns, example.com, *.com
+	portNumColon := strings.Index(ref, ":")
+	if portNumColon != -1 {
+		_, err := strconv.Atoi(ref[portNumColon+1:])
+		if err != nil {
+			return -1
+		}
+		ref = ref[:portNumColon]
+	}
+
+	if strings.HasSuffix(ref, prefix[1:]) {
+		return strings.LastIndex(origRef, "/")
+	}
+	return -1
```

Comment:
The code looks quite complex.  Can you elaborate why we cannot use the `strings.Index(...)` logic we used before? If we do prefix validation at loading time, we can rely on the format.

Reply:
ack, will do